TNT-32999 - add sessionId tracking for personalization component

artur-ciocanu opened the pull request:
Add user session tracking

## Description

Add user session tracking

## Related Issue

NONE

## Motivation and Context

Personalization component can be used to personalize depending on your in-session activities. Hence we need a way to track user session.

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] All tests pass and I've made any necessary test changes.
- [x] I have run the Sandbox successfully.

Aaronius commented on src/components/Personalization/index.js:
```diff
@@ -60,6 +77,11 @@ const createPersonalization = ({ config, logger }) => {
         );
       },
+      onBeforeDataCollection(payload) {
+        const sessionId = getOrCreateSessionId(cookie);
```
Because `getOrCreateSessionId` uses cookie values and writes to cookies, this probably needs to happen _after_ the user opts in. See https://github.com/adobe/alloy/blob/master/src/components/Identity/index.js#L51 for an example of how this is done.

artur-ciocanu replied:
Thanks, I have adjusted.

Aaronius commented on src/components/Personalization/index.js:
```diff
@@ -10,14 +10,31 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 
 const PAGE_HANDLE = "personalization:page";
+const SESSION_ID_COOKIE = "alloy-session-id";
+const SESSION_ID_TTL = 31 * 60 * 1000;
```
31 minutes? What the reasoning behind that value? Just curious more than anything.

artur-ciocanu replied:
This is what we currently have in Adobe Target, basically if there is 31 mins of inactivity.

Aaronius commented on src/components/Personalization/index.js:
```diff
@@ -10,14 +10,31 @@ OF ANY KIND, either express or implied. See the License for the specific languag
+  }
+
+  // We have to extend cookie lifetime
+  cookie.set(SESSION_ID_COOKIE, sessionId, { domain, expires });
```
I believe `cookie` is the alloy cookie accessor object (found in `src/core/createCookie.js`), which doesn't accept a `domain` or `expires`. Since the alloy cookie is a plethora of data, it all shares the same expiry (and domain), which is `180` days.

artur-ciocanu replied:
I have removed the domain and expires and I have added expires as part of the cookie value.

artur-ciocanu commented:
@Aaronius I have addressed all your comments. Could you please take another look.

I have also have a question, may be it's optIn related by I have run into this scenario in sandbox:
1. Clear cache and cookies
2. Reload the page
3. Session ID in personalization is 1111
4. Reload the page again, now session ID is 2222
5. Reload the page again, session ID is 2222

I am not sure either this my implementation or this is related to sandbox or something else.

Aaronius commented:
Huh. I'll take a look.

Aaronius commented:
It's a problem with the cookie proxy inside `createCookie.js`. I'll fix it in a new PR.

Aaronius commented:
Here's the cookie proxy issue: https://jira.corp.adobe.com/browse/CORE-32658

artur-ciocanu commented:
@Aaronius appreciate! Thanks for review!
